File: lab7/lab7-1.java
import java.util.Arrays;
import java.util.Comparator;

class Item {
    int value;
    int weight;
    double ratio;

    // Item constructor
    public Item(int value, int weight) {
        this.value = value;
        this.weight = weight;
        this.ratio = (double) value / weight;  // Value per unit weight
    }
}

public class FractionalKnapsack {

    // Fractional Knapsack function
    public static double fractionalKnapsack(int capacity, Item[] items) {
        // Бараануудыг үнэ/жингийн харьцаагаар эрэмбэлэх
        Arrays.sort(items, Comparator.comparingDouble(item -> -item.ratio));

        double totalValue = 0.0;  // Үр дүнгийн үнэлгээ
        for (Item item : items) {
            if (capacity <= 0) {
                break;
            }

            // Хэрвээ бараа бүгдийг нь оруулж чадахгүй бол хувааж оруулах
            if (item.weight <= capacity) {
                totalValue += item.value;
                capacity -= item.weight;
            } else {
                totalValue += item.value * ((double) capacity / item.weight);
                capacity = 0;  // Багтаамж дууссан
            }
        }
        return totalValue;
    }

    // Жишээ
    public static void main(String[] args) {
        // Жишээ бараанууд
        Item[] items = {
            new Item(60, 10),  // Бараа 1: үнэ=60, жин=10
            new Item(100, 20),  // Бараа 2: үнэ=100, жин=20
            new Item(120, 30)   // Бараа 3: үнэ=120, жин=30
        };
        int capacity = 50;  // Үүргэвчний багтаамж

        double maxValue = fractionalKnapsack(capacity, items);
        System.out.println("Максимум үнэлгээ: " + maxValue);
    }
}

File: lab7/lab7-1.py
class Item:
    def __init__(self, value, weight):
        self.value = value
        self.weight = weight
        self.ratio = value / weight  # Value per unit weight

# Функц, үүргэвчинд оруулах бараануудын хамгийн их үнэлгээг олох
def fractional_knapsack(capacity, items):
    # Бараануудыг үнэ/жингийн харьцаагаар эрэмбэлнэ
    items.sort(key=lambda x: x.ratio, reverse=True)

    total_value = 0  # Үр дүнгийн үнэлгээ
    for item in items:
        if capacity <= 0:
            break
        # Хэрвээ бараа бүгдийг нь оруулж чадахгүй бол хувааж оруулах
        if item.weight <= capacity:
            total_value += item.value
            capacity -= item.weight
        else:
            total_value += item.value * (capacity / item.weight)
            capacity = 0  # Багтаамж дууссан

    return total_value

# Жишээ
if __name__ == "__main__":
    items = [
        Item(60, 10),  # Бараа 1: үнэ=60, жин=10
        Item(100, 20),  # Бараа 2: үнэ=100, жин=20
        Item(120, 30)   # Бараа 3: үнэ=120, жин=30
    ]
    capacity = 50  # Үүргэвчний багтаамж

    max_value = fractional_knapsack(capacity, items)
    print(f"Максимум үнэлгээ: {max_value}")
